### PR TITLE
[corlib] Removes unused managed string indexer code

### LINKED ISSUE #6398: Optimize implementation of string indexer
## Steps to Reproduce

1. Disassemble mscorlib.dll
2. Observe that string indexer `get_Chars` is 0x23 code size
3. Observe that it's above JIT inline limit
4. Observe that .NETFramework/CoreCLR have the string indexing implemented as an icall

## Current Behavior

It seems that mono generates a lot more ASM code for this than Windows .NET:
https://gist.github.com/Therzok/fdc39e06db53898520a9af702d08b098

The .NET JIT code:
https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgBgARYCMA3ALABQuBhAdAEoCuAdjAJZhw0DCA9mAA6sANogDKiAG6sAxnChlK+IvWZsONAJItEPfuIRTZ8ihSwAmagA4uQ6FAoBvCnmoA2AgBYrACgCUzvCdyAEgifFYmfgYYPABePAAiaAATADMEhWDpAAsIBDwADzi8CKiYAG0cAF0FFwBfCgbyIA

## Expected Behavior

Probably would be better performing. Current marks the method as unsafe code, which might not optimize contents:
https://github.com/mono/mono/blob/master/mcs/class/referencesource/mscorlib/system/string.cs#L675-L691

### On which platforms did you notice this

[x] macOS
[ ] Linux
[ ] Windows

**Version Used**:

Mono JIT compiler version 5.11.0.26 (pull-request-6203/f74a01add57 Tue Dec 12 18:08:19 EST 2017)

### PR DESCRIPTION
**Pull request by marek-safar**

### CODE REVIEW

**Comment by lewurm** [comment]
what's the optimization? smaller IL code size?

**Comment by marek-safar** [comment]
smaller IL and `Unsafe.Add` is/should be an intrinsic

**Comment by migueldeicaza** [comment]
It looks like string.get_Chars is already an runtime intrinsic.

**Comment by lewurm** [comment]
@migueldeicaza is right:
https://github.com/mono/mono/blob/0f9bb15622b649db81974057f66669674ea47923/mono/mini/method-to-ir.c#L4826-L4857

@Therzok how did you determine this is a hot path? Since it's a intrinsic, it shouldn't show up in the profiling information (afaik).

**Comment by migueldeicaza** [comment]
There is an extensive discussion on #vsmac currently, and we determined that BenchmarkDotNet is measuring the wrong piece of code.  We reviewed the code, and we are generally faster.

The only downside of the instrinsic is that it seems to trip the inliner and prevents very simple methods that do strig indexing from being inlined (see the discussion)

**Comment by Therzok** [comment]
Determined via timing the code which tokenizes msbuild conditions in VSMac. Read/PeekChar were hot methods.

`void ReadChar() { nextChar = inputString[position]; }` in one of my tries to optimize that further still was slower than I expected it to be

File: mcs/class/corlib/ReferenceSources/String.cs [hunk 1]
```diff
@@ -758,5 +758,14 @@ unsafe String CreateString (ReadOnlySpan<char> value)
 
 			return result;
 		}
+
+		[IndexerName ("Chars")]
+		public char this [int index] {
+			get {
+				if ((uint)index >= m_stringLength)
```
Comment (jaykrell) [inline mcs/class/corlib/ReferenceSources/String.cs]: Nice -- uint beats int for efficient range checks. :)

**Comment by migueldeicaza** [comment]
Given that we treat the indexer as an intrinsic, should we even keep this PR open?

**Comment by marek-safar** [comment]
Yes, because we support "optimization" mode that disables intrinsic in which case the code is going to be called/used

**Comment by marek-safar** [comment]
build

**Comment by marek-safar** [comment]
build

**Pull request merged**